src/lowering/desugar.ml

```diff
@@ -666,15 +666,28 @@ and block force_unit ds =
   match force_unit, last.it with
   | _, S.ExpD e ->
     (decs prefix, exp e)
-  | false, S.LetD ({it = S.VarP x; _}, e, None) ->
+  | false, S.LetD ({it = S.VarP x; _}, e, None) -> (* FIXME need this case for Some fail? *)
     (decs ds, varE (var x.it e.note.S.note_typ))
-  | false, S.LetD (p', e', None) ->
-    let x = fresh_var "x" (e'.note.S.note_typ) in
-    (decs prefix @ [letD x (exp e'); letP (pat p') (varE x)], varE x)
-  | false, S.LetD (p', e', Some fail) ->
-    let x = fresh_var "x" (e'.note.S.note_typ) in
-    let e'' = S.{it = SwitchE(e', [{it = {pat = p'; exp = e'}; at = p'.at; note = ()}; {it = {pat = { it = WildP; at = fail.at; note = fail.note.note_typ }; exp = fail}; at = fail.at ; note = ()}]); at = e'.at; note = e'.note } in
-    (decs prefix @ [letD x (exp e''); letP (pat p') (varE x)], varE x)
+  | false, S.LetD (p, e, None) ->
+    let x = fresh_var "x" (e.note.S.note_typ) in
+    (decs prefix @ [letD x (exp e); letP (pat p) (varE x)], varE x)
+  | false, S.LetD (p, e, Some f) ->
+    let x = fresh_var "x" (e.note.S.note_typ) in
+    let p' = pat p in
+    let e' = exp e in
+    let f' = exp f in
+    let switchE = 
+      I.{ 
+        it = I.SwitchE(
```

Comment:
Why do you prefer `I.SwitchE`?

Reply:
I was keeping in line with the other cases in `block` and `dec'` lowering directly to IR. Seemed less hacky / convoluted to go straight to IR instead of rewriting to another source expression (especially since there wasn't a lot of complexity to do it).